File: client/javascript/src/prsim/network.ts
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

import {mpi} from "@/macro/mpi"
import {spi} from "@/macro/spi"
import {Environ} from "@/kinds/environ";
import {index} from "@/macro/idx";
import {Route} from "@/kinds/route";
import {Page, Paging} from "@/kinds/page";
import {Versions} from "@/kinds/versions";
import {Institution} from "@/kinds/institution";
import type {Context} from "@/prsim/context";
import {Status} from "@/cause";

/**
 * @author coyzeng@gmail.com
 */
@spi("mesh")
export abstract class Network {

    /**
     * Get the meth network environment fixed information.
     *
     * @return Fixed system information.
     */
    @mpi("mesh.net.environ", Environ)
    getEnviron(ctx: Context): Promise<Environ> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Check the mesh network is accessible.
     *
     * @return true is accessible.
     */
    @mpi("mesh.net.accessible", Boolean)
    accessible(ctx: Context, @index(0, 'route') route: Route): Promise<boolean> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Refresh the routes to mesh network.
     */
    @mpi("mesh.net.refresh")
    refresh(ctx: Context, @index(0, 'routes') routes: Route[]): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * getRoute the network edge route.
     *
     * @return edge routes
     */
    @mpi("mesh.net.edge", Route)
    getRoute(ctx: Context, @index(0, 'node_id') node_id: string): Promise<Route> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * getRoutes the network edge routes.
     *
     * @return edge routes
     */
    @mpi("mesh.net.edges", [Array, Route])
    getRoutes(ctx: Context): Promise<Route[]> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * GetNetDomain the network domains.
     *
     * @return net domains
     */
    @mpi("mesh.net.domains", [Array, Route])
    getDomains(ctx: Context): Promise<Route[]> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Put the network domains.
     */
    @mpi("mesh.net.resolve")
    putDomains(ctx: Context, @index(0, 'domains') domains: Route[]): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Weave the network.
     */
    @mpi("mesh.net.weave")
    weave(ctx: Context, @index(0, 'route') route: Route): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Acknowledge the network.
     */
    @mpi("mesh.net.ack")
    ack(ctx: Context, @index(0, 'route') route: Route): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Disable the network.
     */
    @mpi("mesh.net.disable")
    disable(ctx: Context, @index(0, 'node_id') node_id: string): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Enable the network.
     */
    @mpi("mesh.net.enable")
    enable(ctx: Context, @index(0, 'node_id') node_id: string): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Index the network edges.
     */
    @mpi("mesh.net.index", [Page, Route])
    index(ctx: Context, @index(0, 'index') index: Paging): Promise<Page<Route>> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Network environment version.
     *
     * @return version.
     */
    @mpi("mesh.net.version", Versions)
    version(ctx: Context, @index(0, 'node_id') node_id: string): Promise<Versions> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Network institutions.
     */
    @mpi("mesh.net.instx", [Page, Institution])
    instx(ctx: Context, @index(0, 'index') index: Paging): Promise<Page<Institution>> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Network institutions.
     */
    @mpi("mesh.net.instr")
    instr(ctx: Context, @index(0, 'institutions') institutions: Institution[]): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Network form alliance.
     */
    @mpi("mesh.net.ally")
    ally(ctx: Context, @index(0, 'node_ids') node_ids: string[]): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Network quit alliance.
     */
    @mpi("mesh.net.disband")
    disband(ctx: Context, @index(0, 'node_ids') node_ids: string[]): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Network feature assert.
     */
    @mpi("mesh.net.assert", Boolean)
    assert(ctx: Context, @index(0, "feature") feature: string, @index(1, 'node_ids') node_ids: string[]): Promise<boolean> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

}

File: client/golang/prsim/network.go
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package prsim

import (
	"context"
	"github.com/opendatav/mesh/client/golang/macro"
	"github.com/opendatav/mesh/client/golang/types"
)

const (
	AutoDomain = "auto"
	ManuDomain = "manu"
)

var INetwork = (*Network)(nil)

var (
	NetworkRouteRefresh = &macro.Btt{Topic: "mesh.net.route.refresh", Code: "*"}
	RoutePeriodRefresh  = &macro.Btt{Topic: "mesh.plugin.proxy.dynamic.refresh", Code: "*"}
)

// Network spi
// @MPI(macro.MeshMPI)
// @SPI(macro.MeshSPI)
type Network interface {

	// GetEnviron Get the meth network environment fixed information.
	// @MPI("mesh.net.environ")
	GetEnviron(ctx context.Context) (*types.Environ, error)

	// Accessible Check the mesh network is accessible.
	// @MPI("mesh.net.accessible")
	Accessible(ctx context.Context, route *types.Route) (bool, error)

	// Refresh the routes to mesh network.
	// @MPI("mesh.net.refresh")
	Refresh(ctx context.Context, routes []*types.Route) error

	// GetRoute the network edge route.
	// @MPI("mesh.net.edge")
	GetRoute(ctx context.Context, nodeId string) (*types.Route, error)

	// GetRoutes the network edge routes.
	// @MPI("mesh.net.edges")
	GetRoutes(ctx context.Context) ([]*types.Route, error)

	// GetDomains the network domains.
	// @MPI("mesh.net.domains")
	GetDomains(ctx context.Context, kind string) ([]*types.Domain, error)

	// PutDomains the network domains.
	// @MPI("mesh.net.resolve")
	PutDomains(ctx context.Context, kind string, domains []*types.Domain) error

	// Weave the network.
	// @MPI("mesh.net.weave")
	Weave(ctx context.Context, route *types.Route) error

	// Ack the network.
	// @MPI("mesh.net.ack")
	Ack(ctx context.Context, route *types.Route) error

	// Disable the network
	// @MPI("mesh.net.disable")
	Disable(ctx context.Context, nodeId string) error

	// Enable the network
	// @MPI("mesh.net.enable")
	Enable(ctx context.Context, nodeId string) error

	// Index the network edges
	// @MPI("mesh.net.index")
	Index(ctx context.Context, index *types.Paging) (*types.Page[*types.Route], error)

	// Version
	// Network environment version.
	// @MPI("mesh.net.version")
	Version(ctx context.Context, nodeId string) (*types.Versions, error)

	// Instx
	// Network institutions.
	// @MPI("mesh.net.instx")
	Instx(ctx context.Context, index *types.Paging) (*types.Page[*types.Institution], error)

	// Instr
	// Network institutions.
	// @MPI("mesh.net.instr")
	Instr(ctx context.Context, institutions []*types.Institution) error

	// Ally
	// Network form alliance.
	// @MPI("mesh.net.ally")
	Ally(ctx context.Context, nodeIds []string) error

	// Disband
	// Network quit alliance.
	// @MPI("mesh.net.disband")
	Disband(ctx context.Context, nodeIds []string) error

	// Assert
	// Network feature assert.
	// @MPI("mesh.net.assert")
	Assert(ctx context.Context, feature string, nodeIds []string) (bool, error)
}
